presto-main/src/main/java/io/prestosql/execution/resourcegroups/InternalResourceGroup.java

```diff
@@ -646,87 +646,116 @@ private void updateEligibility()
-                return;
-            }
-            // Only count the CPU time if the query succeeded, or the failure was the fault of the user
-            if (!query.getErrorCode().isPresent() || query.getErrorCode().get().getType() == USER_ERROR) {
-                InternalResourceGroup group = this;
-                while (group != null) {
-                    group.cpuUsageMillis = saturatedAdd(group.cpuUsageMillis, query.getTotalCpuTime().toMillis());
-                    group = group.parent.orElse(null);
-                }
-            }
-            if (runningQueries.contains(query)) {
-                runningQueries.remove(query);
+            // Query is queued
+            if (queuedQueries.contains(query)) {
+                queuedQueries.remove(query);
                 InternalResourceGroup group = this;
                 while (group.parent.isPresent()) {
-                    group.parent.get().descendantRunningQueries--;
+                    group.parent.get().descendantQueuedQueries--;
                     group = group.parent.get();
                 }
+                updateEligibilityOnPathToRoot();
+                return;
             }
-            else {
-                queuedQueries.remove(query);
+
+            ResourceUsage lastUsage = runningQueries.get(query);
+            // Query is running
+            if (lastUsage != null) {
+                // CPU is measured cumulatively (i.e. total CPU used until this moment by the query). Memory is measured
+                // instantaneously (how much memory the query is using at this moment). At query completion, memory usage
+                // drops to zero. CPU usage is extracted from query execution.
+                ResourceUsage finalUsage = new ResourceUsage(
+                        query.getTotalCpuTime().toMillis(),
+                        0L);
+                ResourceUsage delta = finalUsage.subtract(lastUsage);
+
+                runningQueries.remove(query);
+
+                // Update usage statistics up to the root
                 InternalResourceGroup group = this;
-                while (group.parent.isPresent()) {
-                    group.parent.get().descendantQueuedQueries--;
-                    group = group.parent.get();
+                while (group != null) {
+                    group.cachedResourceUsage = group.cachedResourceUsage.add(delta);
+                    InternalResourceGroup parent = group.parent.orElse(null);
+                    if (parent != null) {
+                        parent.descendantRunningQueries--;
+                        if (parent.descendantRunningQueries == 0) {
+                            parent.dirtySubGroups.remove(group);
+                        }
+                    }
+                    group = parent;
                 }
+
+                updateEligibilityOnPathToRoot();
+                return;
             }
-            updateEligibility();
+
+            // Query was already cleaned up
```

Comment:
This code would be easier to understand if you did an early exit instead.

Reply:
hmm, I did this refactoring to minimize the number of lookups in runningQueries and queuedQueries  ( https://github.com/prestosql/presto/pull/1128#discussion_r307529092 ). But if this is difficult to read, I can revert it.